### PR TITLE
[ASTGen] Refactor bridgedSourceLoc to SyntaxProtocol

### LINKED ISSUE #68354: `bridgedSourceLoc` should be a member on `SyntaxProtocol`
We have changed the paradigm to prefer passing `ASTGenVisitor` as a parameter to eg. `Collection.map(in:)` rather than defining a function on `ASTGenVisitor` that takes an argument because it puts the subject of the operation at the start, which reads nicer.

To stick with that paradigm, `ASTGenVisitor.bridgedSourceLoc(for node: (some SyntaxProtocol)?)` should become

```swift
extension SyntaxProtocol {
  func bridgedSourceLoc(in astgen: ASTGenVisitor) { … }
}
```

### PR DESCRIPTION
**Pull request by mininny**
Draft PR to resolve https://github.com/apple/swift/issues/68354

### CODE REVIEW

**Comment by ahoppen** [comment]
Looks good to me but let’s put this on hold until https://github.com/apple/swift/pull/67111 is merged because that’s by far the larger PR and it’s easier to rebase this one on top of https://github.com/apple/swift/pull/67111.

**Comment by ahoppen** [comment]
@mininny All occurrences that currently call https://github.com/apple/swift/blob/44a4d6998cd930747214b64d9d25fed1b3ff43f6/lib/ASTGen/Sources/ASTGen/Bridge.swift#L6-L14 should be migrated to `node.bridgedSourceLoc(in:)`. Could you do that as well?

**Comment by mininny** [comment]
> @mininny All occurrences that currently call
> 
> https://github.com/apple/swift/blob/44a4d6998cd930747214b64d9d25fed1b3ff43f6/lib/ASTGen/Sources/ASTGen/Bridge.swift#L6-L14
> 
> should be migrated to `node.bridgedSourceLoc(in:)`. Could you do that as well?

Yep definitely. I see 67111 was just merged. I'll work on this soon.

**Comment by mininny** [comment]
@ahoppen I've converted all of `ASTGenVisitor.bridgedSourceLoc` to `SyntaxProtocol.bridgedSourceLoc` and its optional counterpart when needed.

File: lib/ASTGen/Sources/ASTGen/Bridge.swift [hunk 1]
```diff
@@ -56,6 +44,30 @@ extension String {
...
+  /// Obtains the bridged start location of the node in the source buffer provided by `astgen`.
+  ///
+  /// - Parameter astgen: The visitor providing the `ASTContext` and source buffer.
+  @inline(__always)
+  func bridgedSourceLoc(in astgen: ASTGenVisitor) -> BridgedSourceLoc { 
+    return BridgedSourceLoc(at: self.positionAfterSkippingLeadingTrivia, in: astgen.base)
+  }
+}
+
+extension Optional where Wrapped: SyntaxProtocol {
```
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Bridge.swift:57]: You can shorten this to

```suggestion
extension Optional<SyntaxProtocol> { 
```

**Review by ahoppen — COMMENTED** [review]
Nice. One minor comment, otherwise LGTM.

File: lib/ASTGen/Sources/ASTGen/Bridge.swift (hunk 1 above)
Comment (mininny) [reply lib/ASTGen/Sources/ASTGen/Bridge.swift]: Thanks for the review! I changed it to `Optional<SyntaxProtocol>`

Comment (AnthonyLatsis) [reply lib/ASTGen/Sources/ASTGen/Bridge.swift]: `extension Optional<SyntaxProtocol>` is sugar for `extension Optional where Wrapped == SyntaxProtocol`, not `Wrapped: SyntaxProtocol`.

File: lib/ASTGen/Sources/ASTGen/Bridge.swift [hunk 2]
```diff
@@ -56,6 +44,30 @@ extension String {
   }
 }
 
+extension SyntaxProtocol {
+  /// Obtains the bridged start location of the node in the source buffer provided by `astgen`.
+  ///
+  /// - Parameter astgen: The visitor providing the `ASTContext` and source buffer.
```
Comment (AnthonyLatsis) [inline lib/ASTGen/Sources/ASTGen/Bridge.swift:50]: The AST context is irrelevant to constructing source locations. Could we mention only the source buffer?

File: lib/ASTGen/Sources/ASTGen/Bridge.swift [hunk 3]
```diff
@@ -56,6 +44,30 @@ extension String {
   }
 }
 
+extension SyntaxProtocol {
+  /// Obtains the bridged start location of the node in the source buffer provided by `astgen`.
```
Comment (AnthonyLatsis) [inline lib/ASTGen/Sources/ASTGen/Bridge.swift:48]: Could you please state that the location excludes leading trivia, as the old comment did?

File: lib/ASTGen/Sources/ASTGen/Bridge.swift (hunk 1 above)
Comment (mininny) [reply lib/ASTGen/Sources/ASTGen/Bridge.swift:57]: Thanks guys, I've reverted it back to `where Wrapped: SyntaxProtocol`

File: lib/ASTGen/Sources/ASTGen/Bridge.swift (hunk 2 above)
Comment (mininny) [reply lib/ASTGen/Sources/ASTGen/Bridge.swift]: Thanks, yep!

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by mininny** [comment]
It seems like the tests succeeded. Could someone merge this?
